kore/src/Kore/Log/DebugRewriteTrace.hs

```diff
@@ -0,0 +1,267 @@
+    debugInitialClaim,
+    debugInitialPattern,
+    debugFinalPatterns,
+    debugRewriteTrace,
+    rewriteTraceLogger,
+) where
+
+import Data.ByteString (
+    ByteString,
+ )
+import Data.Text (
+    Text,
+ )
+import Data.Text.Encoding (
+    decodeUtf8,
+ )
+import Data.Yaml (
+    ToJSON,
+    Value (..),
+    array,
+    encode,
+    object,
+    toJSON,
+    (.=),
+ )
+import Kore.Attribute.UniqueId (
+    UniqueId (..),
+ )
+import Kore.Internal.Conditional qualified as Conditional
+import Kore.Internal.MultiOr (
+    MultiOr,
+ )
+import Kore.Internal.Pattern (
+    Pattern,
+ )
+import Kore.Internal.Predicate (
+    unparseWithSort,
+ )
+import Kore.Internal.Substitution (
+    Substitution,
+    assignedTerm,
+    assignedVariable,
+    unwrap,
+ )
+import Kore.Internal.Substitution qualified as Substitution
+import Kore.Internal.TermLike qualified as TermLike
+import Kore.Internal.TermLike.TermLike (
+    TermLike,
+ )
+import Kore.Internal.Variable (
+    VariableName,
+    toVariableName,
+ )
+import Kore.Rewrite.Result qualified as Result
+import Kore.Rewrite.RewritingVariable
+import Kore.Rewrite.RulePattern (
+    UnifyingRuleVariable,
+ )
+import Kore.Rewrite.Step (
+    Results,
+ )
+import Kore.Unparser (
+    Unparse,
+    unparse,
+ )
+import Log
+import Prelude.Kore
+import Pretty (
+    Pretty (..),
+    layoutOneLine,
+    renderText,
+ )
+
+data DebugInitialClaim = DebugInitialClaim UniqueId (TermLike VariableName)
+    deriving stock (Show)
+
+newtype DebugInitialPattern = DebugInitialPattern (TermLike VariableName)
+    deriving stock (Show)
+
+newtype DebugFinalPatterns = DebugFinalPatterns [Pattern VariableName]
+    deriving stock (Show)
+
+data RewriteResult = RewriteResult
+    { ruleId :: UniqueId
+    , substitution :: Substitution VariableName
+    , results :: [Pattern VariableName]
+    }
+    deriving stock (Show)
+
+data DebugRewriteTrace = DebugRewriteTrace
+    { initialPattern :: Pattern VariableName
+    , rewriteResults :: [RewriteResult]
+    , remainders :: [Pattern VariableName]
+    }
+    deriving stock (Show)
+
+instance ToJSON (TermLike VariableName) where
+    toJSON = toJSON . unparseOneLine
+
+instance ToJSON (Pattern VariableName) where
+    toJSON Conditional.Conditional{term, predicate, substitution} =
+        object
+            [ "term" .= term
+            , "constraint" .= encodedConstraint
+            , "substitution" .= substitution
+            ]
+      where
+        sort = TermLike.termLikeSort term
+        encodedConstraint = (renderText $ layoutOneLine $ unparseWithSort sort predicate) :: Text
+
+instance ToJSON (Substitution VariableName) where
+    toJSON substitution =
+        array (encodeKV <$> unwrap substitution)
+      where
+        encodeKV assignment =
+            object
+                [ "key" .= unparseOneLine (assignedVariable assignment)
+                , "value" .= unparseOneLine (assignedTerm assignment)
+                ]
```

Comment:
Probably we should not define these orphans here, they might cause us trouble later if at any point we want _actual_ `ToJson` instances (which are not just strings) for  internal types like `TermLike` or `Pattern`.
We can instead define stand-alone functions "renderX :: X -> Value"  and use those explicitly in the instances for `Debug[Initial|Final]...` below.

Reply:
good point! removed now.